src/librustc_mir/hir.rs

```diff
@@ -0,0 +1,364 @@
+    },
+}
+
+// The HIR trait implementor translates their expressions (`H::Expr`)
+// into instances of this `Expr` enum. This translation can be done
+// basically as lazilly or as eagerly as desired: every recursive
+// reference to an expression in this enum is an `ExprRef<H>`, which
+// may in turn be another instance of this enum (boxed), or else an
+// untranslated `H::Expr`. Note that instances of `Expr` are very
+// shortlived. They are created by `HIR::to_expr`, analyzed and
+// converted into MIR, and then discarded.
+//
+// If you compare `Expr` to the full compiler AST, you will see it is
+// a good bit simpler. In fact, a number of the more straight-forward
+// MIR simplifications are already done in the impl of `HIR`. For
+// example, method calls and overloaded operators are absent: they are
+// expected to be converted into `Expr::Call` instances.
+#[derive(Clone, Debug)]
+pub struct Expr<H:HIR> {
+    // type of this expression
+    pub ty: H::Ty,
+
+    // lifetime of this expression if it should be spilled into a
+    // temporary; should be None unless in a constant contenxt
+    pub temp_lifetime: Option<H::CodeExtent>,
+
+    // span of the expression in the source
+    pub span: H::Span,
+
+    // kind of expression
+    pub kind: ExprKind<H>,
+}
+
+#[derive(Clone, Debug)]
+pub enum ExprKind<H:HIR> {
+    Scope { extent: H::CodeExtent, value: ExprRef<H> },
+    Paren { arg: ExprRef<H> }, // ugh. should be able to remove this!
+    Box { place: Option<ExprRef<H>>, value: ExprRef<H> },
+    Call { fun: ExprRef<H>, args: Vec<ExprRef<H>> },
+    Deref { arg: ExprRef<H> }, // NOT overloaded!
+    Binary { op: BinOp, lhs: ExprRef<H>, rhs: ExprRef<H> }, // NOT overloaded!
+    LogicalOp { op: LogicalOp, lhs: ExprRef<H>, rhs: ExprRef<H> },
+    Unary { op: UnOp, arg: ExprRef<H> }, // NOT overloaded!
+    Cast { source: ExprRef<H> },
+    ReifyFnPointer { source: ExprRef<H> },
+    UnsafeFnPointer { source: ExprRef<H> },
+    Unsize { source: ExprRef<H> },
+    If { condition: ExprRef<H>, then: ExprRef<H>, otherwise: Option<ExprRef<H>> },
+    Loop { condition: Option<ExprRef<H>>, body: ExprRef<H>, },
+    Match { discriminant: ExprRef<H>, arms: Vec<Arm<H>> },
+    Block { body: H::Block },
+    Assign { lhs: ExprRef<H>, rhs: ExprRef<H> },
+    AssignOp { op: BinOp, lhs: ExprRef<H>, rhs: ExprRef<H> },
+    Field { lhs: ExprRef<H>, name: Field<H> },
+    Index { lhs: ExprRef<H>, index: ExprRef<H> },
+    VarRef { id: H::VarId },
+    SelfRef, // first argument, used for self in a closure
+    StaticRef { id: H::DefId },
+    Borrow { region: H::Region, borrow_kind: BorrowKind, arg: ExprRef<H> },
+    Break { label: Option<H::CodeExtent> },
+    Continue { label: Option<H::CodeExtent> },
+    Return { value: Option<ExprRef<H>> },
+    Repeat { value: ExprRef<H>, count: ExprRef<H> },
+    Vec { fields: Vec<ExprRef<H>> },
+    Tuple { fields: Vec<ExprRef<H>> },
+    Adt { adt_def: H::AdtDef,
+          variant_index: usize,
```

Comment:
@arielb1 

> why don't you store the variant itself?

Because it was inconvenient to get from the variant to the containing enum and recompute its index, I think. Maybe I should just add those two bits of information to the VariantDef though.